relay-kafka/src/limits.rs

```diff
@@ -0,0 +1,103 @@
+use std::cmp;
+use std::collections::HashMap;
+use std::fmt;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::time::{Duration, Instant};
+
+use parking_lot::RwLock;
+
+use crate::debounced::Debounced;
+
+type Key = [u8; 16];
+
+/// A really basic rate limiter to protect downstream systems from partition imbalance on ingestion
+/// topics.
+///
+/// The rate limiter is not used to drop messages, but rather to have them be randomly partitioned
+/// (instead of by trace ID) to reduce the partition imbalance again. If the topic is semantically
+/// partitioned this almost certainly will have product impact.
+///
+/// A partition key's message rate is measured in a single window. If the span rate ever exceeds this rate
+/// (`limit_per_window / window_size`), the partition key is limited for up to a duration of
+/// `window_size`.
+#[derive(Clone)]
+pub struct KafkaRateLimits {
+    limit_per_window: u64,
+    map: Arc<RwLock<HashMap<Key, AtomicU64>>>,
```

Comment:
Should use `hashbrown` here, `ahash` is just a free ~10% performance, for a high throughput map like this, it actually matters.